### PR TITLE
Comparing dates correctly

### PR DESCRIPTION
**Pull request by sneakybueno**
Added unit tests to ensure current location is set correctly

### CODE REVIEW

File: FSQLocationBrokerTests/FSQCurrentLocationTests.m [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+- (void)testCurrentLocationWithReverseDates {
+    NSComparisonResult result;
+
+    NSArray<CLLocation *> *locations = @[self.location3, self.location2, self.location1];
+    
+    [self.locationBroker locationManager:self.locationManger didUpdateLocations:locations];
+    
+    result = [self.locationBroker.currentLocation.timestamp compare:self.location3.timestamp];
+    XCTAssert(result == NSOrderedSame, @"expected most recent location");
```
Comment (samgro) [inline FSQLocationBrokerTests/FSQCurrentLocationTests.m]: Can you use `XCTAssertEqual` here? Seems like a more standard/concise way to assert these comparisons.

File: FSQLocationBrokerTests/FSQCurrentLocationTests.m [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+- (void)testCurrentLocationWithInOrderDatesMultipleCalls {
```
Comment (samgro) [inline FSQLocationBrokerTests/FSQCurrentLocationTests.m]: It's not clear exactly what you're testing here. If you can boil the edge case down to 2 calls instead of 3, and more clearly explain what the expected behavior is either in the test name, assertion message, or a comment above the method, that would be helpful. Assertion messages can make it really easy to tell what broke - i.e. "New location with older timestamp should not overwrite existing currentLocation with newer timestamp"

File: FSQLocationBrokerTests/FSQCurrentLocationTests.m [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    
+    NSArray<CLLocation *> *locations1 = @[self.location1];
+    
+    [self.locationBroker locationManager:self.locationManger didUpdateLocations:locations1];
+    
+    result = [self.locationBroker.currentLocation.timestamp compare:self.location3.timestamp];
+    XCTAssert(result == NSOrderedSame, @"expected most recent location");
+}
+
+- (void)testCurrentLocationWithSameTimestamp {
```
Comment (samgro) [inline FSQLocationBrokerTests/FSQCurrentLocationTests.m]: Would this test breaking be a sign of a bug? Does it matter which location wins if there's a timestamp tie?

Comment (sneakybueno) [inline FSQLocationBrokerTests/FSQCurrentLocationTests.m]: Yes, it's better to have a deterministic way of knowing what current location will be when timestamps are equal. I added a comment that stating first location processed wins.

File: FSQLocationBrokerTests/FSQCurrentLocationTests.m (hunk 1 above)
Comment (sneakybueno) [inline FSQLocationBrokerTests/FSQCurrentLocationTests.m]: XCTAssertEqual does not work with NSDate or CLLocation objects. The other problem here is that both `currentLocation` and `timestamp` on CLLocation are both copy properties meaning we can not use pointer equality either. 

This was the quickest way to test we are getting expected results, the best alternative in my mind is adding a category that implements our own `isEqual:` method on CLLocation where we compare individual property values. Thoughts?

Comment (samgro) [inline FSQLocationBrokerTests/FSQCurrentLocationTests.m]: Oh right, ugh. I've used Expecta before (https://github.com/specta/expecta) but don't think we should add a dependency to this open source project. I would vote for adding our own XCTAssertEqualDate and XCTAssertEqualCoordinate methods...

**Comment by sneakybueno** [comment]
@samgro Added testing helper class and macros, the tests are much more readable with those additions.

@beilizhang Can you look into adding support for running unit tests when a PR wants to merge into master?

**Comment by zackzachariah** [comment]
If the unit tests are available as a bitrise.io workflow, I believe that integration for running Unit Tests is trivial.

File: FSQLocationBroker/FSQLocationBroker.m [hunk 4]
```diff
@@ -516,13 +516,13 @@ - (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray
     BOOL isBackgrounded = applicationIsBackgrounded();
     CLLocation *mostRecentLocation = [locations firstObject];
     for (CLLocation *location in locations) {
-        if ([mostRecentLocation.timestamp laterDate:location.timestamp] == location.timestamp) {
+        if ([mostRecentLocation.timestamp compare:location.timestamp] == NSOrderedAscending) {
```
Comment (zackzachariah) [inline FSQLocationBroker/FSQLocationBroker.m]: This will always be false if `mostRecentLocation` is `nil` - I think it would be better to reverse it (i.e. `if ([location.timestamp compare:mostRecentLocation.timestamp] == NSOrderedDescending) {`) so that it will also pass if the incoming `location` is defined and the `mostRecentLocation` isn't.

File: FSQLocationBroker/FSQLocationBroker.m [hunk 5]
```diff
@@ -516,13 +516,13 @@ - (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray
...
             mostRecentLocation = location;
         }
     }
     
     // only update if currentLocation is not set or if mostRecentLocation happened after currentLocation
-    if (!self.currentLocation || [self.currentLocation.timestamp laterDate:mostRecentLocation.timestamp] == mostRecentLocation.timestamp) {
+    if (!self.currentLocation || [self.currentLocation.timestamp compare:mostRecentLocation.timestamp] == NSOrderedAscending) {
```
Comment (zackzachariah) [inline FSQLocationBroker/FSQLocationBroker.m]: ditto about reversing the date check so that it incorporates the `nil` check for you.

File: FSQLocationBroker/FSQLocationBroker.m (hunk 4 above)
Comment (zackzachariah) [inline FSQLocationBroker/FSQLocationBroker.m]: ABORT I AM WRONG

Comment (zackzachariah) [inline FSQLocationBroker/FSQLocationBroker.m]: From the Apple docs:
```
This value must not be nil. If the value is nil, the behavior is undefined and may change in future versions of macOS.
```

Comment (zackzachariah) [inline FSQLocationBroker/FSQLocationBroker.m]: So this is fine, but you should also add the same nil check from below of `!mostRecentLocation ||` (and maybe even still reverse it so that you don't pass if the new `location` is `nil`, and maybe add tests that `nil` values don't mess everything up...)

Comment (sneakybueno) [inline FSQLocationBroker/FSQLocationBroker.m]: So after reading all of these things it sounds like the case we are trying to handle is when `locationManager:didUpdateLocations` contains an array when `locations.count == 0` is that correct? 

I am not sure that will ever happen, but wouldn't it make more sense to just exit the function if it is called with an empty locations array?

**Comment by beilizhang** [comment]
LGTM

File: FSQLocationBroker/FSQLocationBroker.m (hunk 4 above)
Comment (zackzachariah) [inline FSQLocationBroker/FSQLocationBroker.m]: OOooooOOOOh - I hadn't expanded the context to realize that 517-522 are just finding the most recent location from `locations` - yes then nevermind, the check would be unnecessary anyways because we'll never get into the for loop if `locations.count == 0`

File: FSQLocationBroker/FSQLocationBroker.m (hunk 5 above)
Comment (zackzachariah) [inline FSQLocationBroker/FSQLocationBroker.m]: Technically this could still have undefined behavior if `locations` is `nil` or `@[]` because then this could wind up comparing a defined `NSDate` to `nil`, but I agree the risk is very minimal.

File: FSQLocationBrokerTests/FSQTestingHelper.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+//  Copyright © 2017 Foursquare. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+#import <CoreLocation/CoreLocation.h>
+
+#pragma mark Date Equality Macros
+
+#define FSQAssertEqualDates(date1, date2) \
```
Comment (samgro) [inline FSQLocationBrokerTests/FSQTestingHelper.h]: Do these need to be macros or can they be functions? I always prefer functions whenever possible, much easier to debug and harder to shoot yourself in the foot.

File: FSQLocationBrokerTests/FSQCurrentLocationTests.m [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    self.location3 = [[CLLocation alloc] initWithCoordinate:[FSQTestingHelper foursquareNYCCoordinates] altitude:0 horizontalAccuracy:5 verticalAccuracy:0 timestamp:afterFutureDate];
+}
+
+- (void)tearDown {
+    self.locationBroker = nil;
+
+    [super tearDown];
+}
+
+- (void)testDateEqualityMacros {
```
Comment (samgro) [inline FSQLocationBrokerTests/FSQCurrentLocationTests.m]: These probably belong in their own test class? /shrug

File: FSQLocationBrokerTests/FSQTestingHelper.h (hunk 6 above)
Comment (samgro) [inline FSQLocationBrokerTests/FSQTestingHelper.h]: I was thinking of just a C function. If XCTAssert is implemented as a macro, I'm ok with keeping it that way. I'm a fan of using C functions for things like this though - it doesn't need to belong to a class or object, and matches the style of the built-in XCTAssert macro...
